[Wallet] Fee fixes/refactor

jeanregisser opened the pull request:
This PR is relatively big, sorry for this 🙏 
Thanks @jmrossy for the discussions and early feedback that helped shape this PR!

### Description

**The problem**:
There is a single `suggestedFee` in the redux store used for both send and invite. 
Its value is not updated at appropriate times, or not at all, leading to issues like #36.
Also in `SendAmount`, the fee can be calculated too many times.

**This solution**:
The main idea is to have different types of (cached) estimated fees (right now for `send` and `invite`, later on for `exchange` and `escrow`) and trigger their update very infrequently depending on what flow the user is starting.
Then only when going to the `SendConfirmation` or `ReclaimPaymentConfirmationScreen` screens the actual fee is calculated once. Showing some feedback while it does it.

### Tested

Added/Updated tests.

Updated UI with loading/error feedback:
![Screenshot_1564563361](https://user-images.githubusercontent.com/57791/62292574-ffa24800-b466-11e9-987d-9cd82058b9a6.png)
![Screenshot_1564650868](https://user-images.githubusercontent.com/57791/62292684-3d06d580-b467-11e9-8469-746b5316bdeb.png)
![Screenshot_1564563456_censored(1)](https://user-images.githubusercontent.com/57791/62293130-36c52900-b468-11e9-9e6a-516f3985dae4.jpg)
![Screenshot_1564762306](https://user-images.githubusercontent.com/57791/62384908-e1664600-b553-11e9-84a9-14fd8f00f022.png)
![Screenshot_1564762309](https://user-images.githubusercontent.com/57791/62384916-e5926380-b553-11e9-8bfe-aa0285e11ed2.png)

### Other changes

N/A

### Related issues

- Fixes #36, #83, #113

### Backwards compatibility

Backward compatible.

jeanregisser committed:
[wip] Split the different fee estimates

jeanregisser committed:
Use new fee estimate in SendAmount

jeanregisser committed:
Calculate fee in SendConfirmation screen

jeanregisser committed:
Remove commented out code

jeanregisser committed:
Don't persist fees

jeanregisser committed:
Stop using obsolete suggestedFee in PaymentRequestNotification

jeanregisser committed:
Show loading spinner while calculating fee

jeanregisser committed:
Show fee loading indicator inline

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/refactor-fee-take-two

jeanregisser committed:
Add fee calculation for reclaim from escrow

jeanregisser committed:
Use fee variable already in the scope

jeanregisser committed:
Remove obsolete code related to suggestedFee

jeanregisser committed:
Fix gasCurrency not set when calculating send fee

jeanregisser committed:
Remove unnecessary types

jeanregisser committed:
Fix gasCurrency when calculating reclaim escrow fee

jeanregisser committed:
Remove unnecessary any type which would actual errors

jeanregisser committed:
Add error state when fee calculation fails

jeanregisser committed:
Remove obsolete fee parameter from SendConfirmation input

jeanregisser committed:
Remove empty block

jeanregisser commented:
I haven't tackled the hardcoded fee in `InviteReview`
https://github.com/celo-org/celo-monorepo/blob/f272bc826af3bf73843f9d06cfe2da181e63cc1c/packages/mobile/src/account/InviteReview.tsx#L64-L65
and in `getInvitationVerificationFee`
https://github.com/celo-org/celo-monorepo/blob/f272bc826af3bf73843f9d06cfe2da181e63cc1c/packages/mobile/src/invite/saga.ts#L57-L60

I want to address those in a subsequent PR.

jmrossy commented on packages/mobile/src/fees/CalculateFee.tsx:
```diff
@@ -0,0 +1,132 @@
+    () => {
+      // Generic error banner
+      if (asyncResult.error) {
+        showErrorFunction(ErrorMessages.CALCULATE_FEE_FAILED, ERROR_BANNER_DURATION)
```
This is a fine approach, just curious if there's a way to have the showError function be used directly here rather than passed in?

jeanregisser replied:
Yes using `react-redux >= 7.1.0` and its support for [hooks](https://react-redux.js.org/api/hooks#usedispatch), I wanted to do that first but held off as I felt this PR was too big already to couple with a major update of `react-redux`

jmrossy replied:
Agreed

jmrossy commented on packages/mobile/src/fees/CalculateFee.tsx:
```diff
@@ -0,0 +1,132 @@
+
+const CalculateInviteFee: FunctionComponent<DispatchProps & InviteProps> = (props) => {
+  const asyncResult = useAsyncShowError(getInvitationVerificationFee, [], props.showError)
+  return props.children(asyncResult)
```
Leave a TODO here for doing this part? Can you create a Github issue to track please :)

jeanregisser replied:
I'm not following, which part?

jmrossy replied:
My bad, misunderstood this line. Nvm

jmrossy commented on packages/mobile/src/fees/CalculateFee.tsx:
```diff
@@ -0,0 +1,132 @@
+
+const CalculateSendFee: FunctionComponent<DispatchProps & SendProps> = (props) => {
+  const asyncResult = useAsyncShowError(
+    async (account: string, recipientAddress: string, amount: BigNumber, comment: string) =>
```
Why is this function async? We don't seem to use await. If getSendFee returns a promise maybe return that directly

jeanregisser replied:
Good catch, `async` is unnecessary here.

jmrossy commented on packages/mobile/src/fees/CalculateFee.tsx:
```diff
@@ -0,0 +1,132 @@
+      return null
+  }
+
+  return null
```
Consider throwing an error here. We have an error boundary set up for the app

jeanregisser replied:
Sounds good!

jmrossy commented on packages/mobile/src/fees/actions.ts:
```diff
@@ -0,0 +1,35 @@
+  RECLAIM_ESCROW = 'reclaim-escrow',
+}
+
+export interface UpdateDefaultFeeAction {
```
The term `Default` is confusing me a bit here. What do you mean by default fee?

jeanregisser replied:
I agree. I think Estimate is appropriate.
What do you think?

jmrossy replied:
Estimate sgtm

jmrossy commented on packages/mobile/src/fees/saga.test.ts:
```diff
@@ -0,0 +1,32 @@
+import { watchUpdateDefaultFee } from 'src/fees/saga'
+import { getInvitationVerificationFee } from 'src/invite/saga'
+
+// jest.mock('src/utils/time', () => ({
```
Remove dead code here

jeanregisser replied:
Yep!

jmrossy commented on packages/mobile/src/fees/saga.ts:
```diff
@@ -0,0 +1,50 @@
+export function* updateDefaultFeeSaga({ feeType }: UpdateDefaultFeeAction) {
+  Logger.debug(TAG + '@updateDefaultFeeSaga', 'updating for ', feeType)
+
+  // TODO: skip fee update if it was calculated recently
```
Yeah I think we should also store a timestamp for each fee type and check that here

jeanregisser replied:
Yes! Do you mind if I do it in another PR?

jmrossy replied:
Another Pr makes sense :)

jmrossy commented on packages/mobile/src/fees/selectors.ts:
```diff
@@ -0,0 +1,47 @@
+import { RootState } from 'src/redux/reducers'
+import { divideByWei } from 'src/utils/formatting'
+
+const getInviteFeeEstimate = (state: RootState) => state.fees.invite.feeInWei
```
For clarity, can we append `InWei` to these names?

jeanregisser replied:
Yes!

jmrossy reviewed (commented):
Very nice!! Just some small comments/questions below

jeanregisser committed:
Remove unnecessary async keyword

jeanregisser committed:
Throw error on unsupported feeType

martinvol commented on packages/mobile/locales/es-AR/global.json:
```diff
@@ -82,5 +82,6 @@
   "qrFailedInvalidAddress":
-    "Código QR no se pudo leer. Razón: la dirección de la cartera no es válida."
+    "Código QR no se pudo leer. Razón: la dirección de la cartera no es válida.",
+  "calculateFeeFailed": "No se pudo calcular la tarifa"
```
"la tarifa" -> "la comisión"

jeanregisser replied:
Thanks @martinvol :+1:

jeanregisser committed:
Rename default fee -> estimate fee

jeanregisser committed:
Group fee estimates in redux so it's more explicit as to what they are

jeanregisser committed:
Better spanish translation

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/refactor-fee-take-two

jeanregisser committed:
Tweak log message

jeanregisser commented:
I addressed all existing comments.

cmcewen commented on packages/mobile/src/fees/CalculateFee.tsx:
```diff
@@ -0,0 +1,133 @@
+import { getInvitationVerificationFee } from 'src/invite/saga'
+import { getSendFee } from 'src/send/saga'
+
+export type CalculateFeeChildren = (asyncResult: UseAsyncReturn<BigNumber, never>) => any
```
Will `React.ReactNode` work here instead of any?

jeanregisser replied:
Let me try.

jeanregisser replied:
Here it is: https://github.com/celo-org/celo-monorepo/commit/fcbc7ab871bf9f43876722957006fa311f4188f6
It caused another type issue, which I worked around with a cast.
Anyway definitely better with `ReactNode` 👍

cmcewen commented on packages/mobile/src/send/SendAmount.tsx:
```diff
@@ -241,6 +233,10 @@ export class SendAmount extends React.PureComponent<Props, State> {
 
     const confirmationInput = this.getConfirmationInput()
+    if (!confirmationInput) {
+      return
```
what could cause a failure case here? should we throw?

jeanregisser replied:
Good catch I don't need this anymore. I had to add this when `fee` was part of `ConfirmationInput`. But I've now removed it.

cmcewen commented on packages/mobile/src/send/SendConfirmation.tsx:
```diff
@@ -176,34 +181,34 @@ class SendConfirmation extends React.Component<Props, State> {
       action: this.onSendButtonClick,
       text: t('send'),
-      disabled: isSending || !userHasEnough || !appConnected,
+      disabled: isSending || !userHasEnough || !appConnected || !!asyncFee.error,
```
dupe logic with below - assign to a variable

jeanregisser replied:
Yep 👍

cmcewen commented on packages/mobile/src/send/SendConfirmation.tsx:
```diff
@@ -176,34 +181,34 @@ class SendConfirmation extends React.Component<Props, State> {
-    )
-    const userHasEnough = amountWithFees.isLessThanOrEqualTo(currentBalance)
+    const fee = asyncFee.result && getFeeDollars(asyncFee.result)
+    const amountWithFee = new BigNumber(numeral(amount).value()).plus(fee || 0)
```
just to be clear - the intention here would be to show the amount the user has already entered while we are calculating the fee, and then update once the fee has been calculated? just wondering if seeing a number jump would be a bad UX, and we should just show loading instead (which is maybe covered by that isLoadingFee?)

jeanregisser replied:
Good point. 
What's happening right now is the send fee is so small it displays as `$0.0000` after loading and the total doesn't change.
The invite fee though is more but since it's [currently hardcoded](https://github.com/celo-org/celo-monorepo/blob/f272bc826af3bf73843f9d06cfe2da181e63cc1c/packages/mobile/src/invite/saga.ts#L57-L60), there's no loading indicator happening.

With the way the UI is done currently, I think the fact that the total can change after the fee has loaded shouldn't be surprising for the user.
But we need to confirm this. If it's a problem we can tweak this.

cmcewen commented on packages/mobile/src/send/SendConfirmation.tsx:
```diff
@@ -248,6 +255,21 @@ class SendConfirmation extends React.Component<Props, State> {
+  render() {
+    const { account } = this.props
+    if (!account) {
+      throw Error('Account is required')
```
Similar question on error handling - when will we ever reach this state?

jeanregisser replied:
This could never happen to our users (unless we mess something and `SendConfirmation` is displayed before an account is created 😸)

Do you have another way to handle this?

jmrossy replied:
We generally don't do much invariant checking in our render functions. And account is something that we can generally rely on being set up properly. I'd recommend just removing this check

jeanregisser replied:
We can't really remove it otherwise TS complains about `account` which can be null.

cmcewen reviewed (commented):
Very nice! Would you feel comfortable walking the team through this one at some point as an optional intro to hooks?

jeanregisser commented:
@cmcewen I'll have a better example of hooks soon which will be short and simple and would convey the advantage of using them.

jeanregisser committed:
Fix children return type

I had to workaround another type issue with this change.
See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18051

jeanregisser committed:
Remove obsolete check

jeanregisser committed:
Use a variable instead of duplicating the logic

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/refactor-fee-take-two

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/refactor-fee-take-two

jeanregisser committed:
Remove code related to fetchReclaimSuggestedFee which is now redundant

jeanregisser committed:
Fix content wrapped by CalculateFee not re-rendered on state change

jeanregisser committed:
Tweak LineItemRow style in reclaim screen

Should probably be extracted into a common component used by both the
reclaim confirmation and send confirmation.

jeanregisser commented:
Latest master with reclaim escrowed has been merged and tested end-to-end successfully! (updated the screenshots in the description).

It's all looking good to me here.

jmrossy commented on packages/mobile/src/fees/CalculateFee.tsx:
```diff
@@ -0,0 +1,135 @@
+      return <CalculateReclaimEscrowFee {...props} />
+  }
+
+  // @ts-ignore
```
Why ts-ignore here?

jeanregisser replied:
TypeScript complains about `feeType` not being present, though it's present on all types of the distinct union. I think I can improve this a little bit and get rid of `@ts-ignore`.

jeanregisser replied:
Hum looks like it doesn't complain anymore. I guess it was a specific condition in a previous iteration that made me add that.

jmrossy commented on packages/mobile/src/fees/CalculateFee.tsx:
```diff
@@ -0,0 +1,135 @@
+  // @ts-ignore
+  throw new Error(`Unsupported feeType: ${props.feeType}`)
+
+  return null
```
no need for return after throw

jeanregisser replied:
Indeed, removing it.

jmrossy commented on packages/mobile/src/fees/saga.ts:
```diff
@@ -0,0 +1,50 @@
+      // TODO
+      break
+    case FeeType.RECLAIM_ESCROW:
+      // TODO
```
Did you mean to leave this as TODO? It looks like you more or less implemented it already

jeanregisser replied:
True. Though we're not yet using the estimate for this fee type, I'll add it for completeness.

jmrossy reviewed (commented):
A few small questions/comments but mostly LGTM

jeanregisser committed:
Remove unnecessary ts-ignore and return

jeanregisser committed:
Add implementation for estimating reclaim escrow fee

jmrossy reviewed (approved):
Ship it!

jeanregisser committed:
Merge branch 'master' into jeanregisser/refactor-fee-take-two

isabellewei commented:
I think this also fixes #112 and #110 ?

jeanregisser commented:
Not directly, the fee for the invite is still using the [hardcoded values](https://github.com/celo-org/celo-monorepo/pull/276#issuecomment-517271997).
